File: lab 5_2.cpp
#include <iostream>   
#include <vector>     
#include <algorithm>  
#include <iomanip>    // Для форматирования вывода (setw, setfill)
#include <sstream>    // Для работы со строками (ostringstream)

using namespace std;  

// Структура, описывающая окно приема
struct Window {
    int total = 0;      // Общее время обработки в этом окне
    vector<string> tickets;  // Список талонов в этом окне
    int num;          // Номер окна
};

// Структура, описывающая посетителя
struct Visitor {
    int time;           // Продолжительность визита 
    string ticket;          // Номер талона
};

// Функция распределения очереди по окнам
vector<Window> distribute(const vector<Visitor>& queue, int num_windows) {
    vector<Window> windows(num_windows);// Создаем вектор окон с заданным количеством
    
    // Инициализируем номера окон
    for (int i = 0; i < num_windows; ++i) {
        windows[i].num = i + 1;
    }

    // Распределяем посетителей по окнам
    for (const auto& visitor : queue) {
        // Находим окно с минимальным текущим временем обработки
        auto window = min_element(windows.begin(), windows.end(),
            [](const Window& a, const Window& b) {
                return a.total < b.total;
            });
        
        // Добавляем время посетителя к выбранному окну
        window->total += visitor.time;
        // Добавляем талон посетителя в это окно
        window->tickets.push_back(visitor.ticket);
    }

    return windows;
}

int main() {
    // Запрос количества окон
    int num_windows;
    cout << ">>> Введите кол-во окон\n<<< ";
    cin >> num_windows;

    // Очередь посетителей
    vector<Visitor> queue;
    // Счетчик для генерации номеров талонов
    int ticket_num = 1;

    // Основной цикл обработки команд
    while (true) {
        string command;
        cout << "<<< ";
        cin >> command;

        // Обработка команды ENQUEUE
        if (command == "ENQUEUE") {
            int enqueue;
            cin >> enqueue;
            
            // Генерация номера талона
            ostringstream num;
            num << "T" << setw(3) << setfill('0') << ticket_num++;
            string ticket = num.str();
            
            // Добавление посетителя в очередь
            queue.push_back({enqueue, ticket});
            cout << ">>> " << ticket << '\n';
        }
        // Обработка команды DISTRIBUTE
        else if (command == "DISTRIBUTE") {
            // Распределение очереди по окнам
            auto windows = distribute(queue, num_windows);
            
            // Сортировка окон по номеру
            sort(windows.begin(), windows.end(),
                [](const Window& a, const Window& b) {
                    return a.num < b.num;
                });

            // Вывод результатов распределения
            for (const auto& window : windows) {
                cout << ">>> Окно " << window.num << " (" << window.total << " минут): ";
                
                // Вывод списка талонов через запятую
                bool first = true;
                for (const auto& ticket : window.tickets) {
                    if (!first) cout << ", ";
                    cout << ticket;
                    first = false;
                }
                cout << '\n';
            }
            break;  // Завершение работы после распределения
        }
        // Обработка неизвестных команд
        else {
            cout << ">>> Неизвестная команда\n";
        }
    }

    return 0;
}

File: lab 5_2.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

// Структура, описывающая окно приема
type Window struct {
	total   int       // Общее время обработки в этом окне
	tickets []string  // Список талонов в этом окне
	num     int       // Номер окна
}

// Структура, описывающая посетителя
type Visitor struct {
	time   int    // Продолжительность визита
	ticket string // Номер талона
}

// Функция распределения очереди по окнам
func distribute(queue []Visitor, numWindows int) []Window {
	windows := make([]Window, numWindows) // Создаем слайс окон с заданным количеством

	// Инициализируем номера окон
	for i := 0; i < numWindows; i++ {
		windows[i].num = i + 1
	}

	// Распределяем посетителей по окнам
	for _, visitor := range queue {
		// Находим окно с минимальным текущим временем обработки
		minIndex := 0
		for i := 1; i < len(windows); i++ {
			if windows[i].total < windows[minIndex].total {
				minIndex = i
			}
		}

		// Добавляем время посетителя к выбранному окну
		windows[minIndex].total += visitor.time
		// Добавляем талон посетителя в это окно
		windows[minIndex].tickets = append(windows[minIndex].tickets, visitor.ticket)
	}

	return windows
}

func main() {
	// Запрос количества окон
	var numWindows int
	fmt.Print(">>> Введите кол-во окон\n<<< ")
	fmt.Scan(&numWindows)

	// Очередь посетителей
	var queue []Visitor
	// Счетчик для генерации номеров талонов
	ticketNum := 1

	// Основной цикл обработки команд
	for {
		var command string
		fmt.Print("<<< ")
		fmt.Scan(&command)

		// Обработка команды ENQUEUE
		if command == "ENQUEUE" {
			var duration int
			fmt.Scan(&duration)

			// Генерация номера талона
			ticket := "T" + fmt.Sprintf("%03d", ticketNum)
			ticketNum++

			// Добавление посетителя в очередь
			queue = append(queue, Visitor{duration, ticket})
			fmt.Println(">>>", ticket)
		} else if command == "DISTRIBUTE" {
			// Обработка команды DISTRIBUTE
			windows := distribute(queue, numWindows)

			// Сортировка окон по номеру
			sort.Slice(windows, func(i, j int) bool {
				return windows[i].num < windows[j].num
			})

			// Вывод результатов распределения
			for _, window := range windows {
				fmt.Printf(">>> Окно %d (%d минут): %s\n", 
					window.num, window.total, strings.Join(window.tickets, ", "))
			}
			break // Завершение работы после распределения
		} else {
			// Обработка неизвестных команд
			fmt.Println(">>> Неизвестная команда")
		}
	}
}
